clients/src/test/java/org/apache/kafka/clients/InFlightRequestsTest.java

```diff
@@ -0,0 +1,188 @@
+    }
+
+
+    @Test
+    public void testLastSent() {
+        NetworkClient.InFlightRequest lastSentRequest = requests.get(requests.size() - 1);
+        assertEquals(lastSentRequest, inFlightRequests.lastSent(lastSentRequest.destination));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testLastSentWithInvalidNode() {
+        inFlightRequests.lastSent("");
+    }
+
+    @Test
+    public void testCompleteLastSent() {
+        NetworkClient.InFlightRequest lastSentRequest = requests.remove(requests.size() - 1);
+        decrementCounter(lastSentRequest.destination);
+        assertEquals(lastSentRequest, inFlightRequests.completeLastSent(lastSentRequest.destination));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testCompleteLastSentWithInvalidNode() {
+        inFlightRequests.completeLastSent("");
+    }
+
+    @Test
+    public void testCanSendMore() {
+    }
+
+    @Test
+    public void testCount() {
+        assertEquals(requests.size(), inFlightRequests.count());
+    }
+
+    @Test
+    public void testIsEmpty() {
+        assertEquals(inFlightRequests.count() == 0, inFlightRequests.isEmpty());
+    }
+
+    @Test
+    public void testCountInSingleNode() {
+        String nodeId = (nodeIdCounter++).toString();
+        int totalRequest = 10;
+        pushRequests(nodeId, totalRequest);
+        assertEquals(totalRequest, inFlightRequests.count(nodeId));
+    }
+
+    @Test
+    public void testIsEmptyInSingleNode()  {
+        String nodeId = (nodeIdCounter++).toString();
+        assertTrue(inFlightRequests.isEmpty(nodeId));
+
+        nodeId = (nodeIdCounter++).toString();
+        int totalRequest = 10;
+        pushRequests(nodeId, totalRequest);
+
+        assertEquals(totalRequest, inFlightRequests.count(nodeId));
+        inFlightRequests.clearAll(nodeId);
+        assertTrue(inFlightRequests.isEmpty(nodeId));
+    }
+
+    @Test
+    public void testClearAll() {
+        String nodeId = (nodeIdCounter++).toString();
+        int totalRequest = 10;
+        pushRequests(nodeId, totalRequest);
+
+        assertEquals(totalRequest, inFlightRequests.count(nodeId));
+        inFlightRequests.clearAll(nodeId);
+        assertEquals(0, inFlightRequests.count(nodeId));
+    }
+
+    @Test
+    public void testGetNodesWithTimedOutRequests() {
+    }
+
+    private void pushRequests(String nodeIdString, int totalRequest) {
+        for (int request = 0; request < totalRequest; request++) {
+            NetworkClient.InFlightRequest inFlightRequest = createInFlightRequest(nodeIdString);
+            requests.add(inFlightRequest);
+            incrementCounter(nodeIdString);
+            inFlightRequests.add(inFlightRequest);
+        }
+    }
+
+    private NetworkClient.InFlightRequest createInFlightRequest(String nodeIdString) {
+        RequestHeader header = new RequestHeader((short) 0, (short) 0, "client", 1);
+        Time time = new SystemTime();
```

Comment:
It will be better to use `MockTime` rather than `SystemTime`. That will make it easier to test timeouts